integration/read_write_mode_test.go

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+//go:build requires_docker
+// +build requires_docker
+
+package integration
+
+import (
+	"math"
+	"testing"
+	"time"
+
+	"github.com/grafana/e2e"
+	e2edb "github.com/grafana/e2e/db"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/mimir/integration/e2emimir"
+)
+
+func TestReadWriteMode(t *testing.T) {
+	s, err := e2e.NewScenario(networkName)
+	require.NoError(t, err)
+	defer s.Close()
+
+	minio := e2edb.NewMinio(9000, blocksBucketName, alertsBucketName, rulerBucketName)
+	require.NoError(t, s.StartAndWaitReady(minio))
+
+	flags := map[string]string{
+		"-common.storage.backend":              "s3",
+		"-common.storage.s3.access-key-id":     e2edb.MinioAccessKey,
+		"-common.storage.s3.secret-access-key": e2edb.MinioSecretKey,
+		"-common.storage.s3.endpoint":          minio.NetworkHTTPEndpoint(),
+		"-common.storage.s3.insecure":          "true",
+		"-blocks-storage.s3.bucket-name":       blocksBucketName,
+		"-alertmanager-storage.s3.bucket-name": alertsBucketName,
+		"-ruler-storage.s3.bucket-name":        rulerBucketName,
+		"-ingester.ring.replication-factor":    "1",
+		"-memberlist.join":                     "mimir-backend-1",
+	}
+
+	readInstance := e2emimir.NewReadInstance("mimir-read-1", flags)
+	writeInstance := e2emimir.NewWriteInstance("mimir-write-1", flags)
+	backendInstance := e2emimir.NewBackendInstance("mimir-backend-1", flags)
+	require.NoError(t, s.StartAndWaitReady(readInstance, writeInstance, backendInstance))
+
+	c, err := e2emimir.NewClient(writeInstance.HTTPEndpoint(), readInstance.HTTPEndpoint(), "", "", "user-1")
+	require.NoError(t, err)
+
+	// Push some data to the cluster.
+	now := time.Now()
+	series, expectedVector := generateSeries("test_series_1", now, prompb.Label{Name: "foo", Value: "bar"})
+
+	res, err := c.Push(series)
+	require.NoError(t, err)
+	require.Equal(t, 200, res.StatusCode)
+
+	// Verify we can read the data we just pushed, both with an instant query and a range query.
+	result, err := c.Query("test_series_1", now)
```

Comment:
This test will only read data from the write instances - do we also want to check that we can query data persisted to the object store as well?

Reply:
Yes, but testing long-term storage is way more annoying, so I would merge this PR as is and then having you re-iterate on it to add querying long-term storage.

Reply:
Cool, I'll add that separately.